src/workerd/io/worker.c++

```diff
@@ -2953,89 +2966,77 @@ void Worker::Actor::Impl::HooksImpl::updateAlarmInMemory(kj::Maybe<kj::Date> new
-    });
-  };
+auto Worker::Actor::getAlarm(kj::Date scheduledTime) -> kj::Maybe<kj::Promise<AlarmResult>> {
+  KJ_IF_MAYBE(runningAlarm, impl->maybeRunningAlarm) {
+    if (runningAlarm->scheduledTime == scheduledTime) {
+      // The running alarm has the same time, we can just wait for it.
+      return runningAlarm->resultPromise.addBranch();
+    }
+  }
 
-  auto makeQueuedAlarm = [&, scheduledTime](auto runningProm) mutable {
-    auto [prom, fulfiller] = kj::newPromiseAndFulfiller<WorkerInterface::AlarmResult>();
-    auto& fulfillerRef = *fulfiller;
-
-    return Impl::Alarm {
-      runningProm.then([runAlarmImpl = kj::mv(runAlarmImpl), &fulfillerRef]() mutable {
-        return runAlarmImpl(fulfillerRef);
-      }),
-      prom.fork(),
-      kj::mv(fulfiller),
-      scheduledTime
-    };
-  };
+  KJ_IF_MAYBE(scheduledAlarm, impl->maybeScheduledAlarm) {
+    if (scheduledAlarm->scheduledTime == scheduledTime) {
+      // The scheduled alarm has the same time, we can just wait for it.
+      return scheduledAlarm->resultPromise.addBranch();
+    }
+  }
 
-  KJ_IF_MAYBE(r, impl->runningAlarm) {
-    if (r->scheduledTime == scheduledTime) {
-      return r->alarm.addBranch();
-    } else KJ_IF_MAYBE(q, r->queuedAlarm) {
-      if (q->scheduledTime == scheduledTime) {
-        return q->alarm.addBranch();
-      } else {
-        // cancel the old invocations
-        q->fulfiller->fulfill(WorkerInterface::AlarmResult {
-          .retry = false,
-          .outcome = EventOutcome::CANCELED
-        });
+  return nullptr;
+}
 
-        // now we can replace the queued alarm with a new one. we exclusiveJoin with the paf promise
-        // to allow for future overwrites
-        return r->queuedAlarm
-          .emplace(makeQueuedAlarm(r->alarm.addBranch().ignoreResult()))
-          .alarm.addBranch();
-      }
-    } else {
-      // there's not a queued alarm already, so we're safe to just go ahead and set it.
-      return r->queuedAlarm
-        .emplace(makeQueuedAlarm(r->alarm.addBranch().ignoreResult()))
-        .alarm.addBranch();
+auto Worker::Actor::scheduleAlarm(kj::Date scheduledTime) -> kj::Promise<ScheduleAlarmResult> {
+  KJ_IF_MAYBE(runningAlarm, impl->maybeRunningAlarm) {
+    if (runningAlarm->scheduledTime == scheduledTime) {
+      // The running alarm has the same time, we can just wait for it.
+      auto result = co_await runningAlarm->resultPromise.addBranch();
+      co_return result;
     }
-  } else {
-    auto [prom, fulfiller] = kj::newPromiseAndFulfiller<WorkerInterface::AlarmResult>();
-    auto& fulfillerRef = *fulfiller;
-    auto& running = impl->runningAlarm.emplace(Impl::RunningAlarm {
-      Impl::Alarm {
-        runAlarmImpl(fulfillerRef),
-        prom.fork(),
-        kj::mv(fulfiller),
-        scheduledTime
-      }
-    });
-    return running.alarm.addBranch();
   }
+
+  KJ_IF_MAYBE(scheduledAlarm, impl->maybeScheduledAlarm) {
+      // We had a previously scheduled alarm, let's cancel it.
```

Comment:
What if it was for the same time? Do we really want to cancel it even in that case or would we be better off just adding a branch to its result promise?

Reply:
So the only caller of `scheduleAlarm()` calls `getAlarm()` in the same synchronous scope, essentially doing what you describe. I'll add the comment in the header!